lib/practice_exercises.rb

```diff
@@ -1,13 +1,55 @@
-
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n^2) because we have an if loop inside an until loop, both depend on n
+# Space Complexity: O(n) because the size that the array takes up depends on its length, ie. n
 def is_palindrome(string)
-  raise NotImplementedError, "Not implemented yet"
+  @string = string
+  # downcase  & strip all the nonalphanumeric characters, and save as a new variable @alphanumeric_only_string
+  @alphanumeric_only_string = @string.downcase.delete('^A-Za-z0-9')
+
+  # convert @alphanumeric_only_string to an array @alphanumeric_array
+  @alphanumeric_array = @alphanumeric_only_string.chars
+
+  # iterate through array and compare first & last, then repeat for next ones
+  # technically we only have to go up to half of the length of the array but then we would have to divide up
+  # the problem in two cases for arrays of even or odd lengths, so this will do
+  i = 0
+  j = @alphanumeric_array.length - 1
+  sum_of_diff = 0
+  until i == @alphanumeric_array.length
+    if @alphanumeric_array[i] == @alphanumeric_array[j]
+      #do nothing
+    else
+      sum_of_diff += 1
+    end
+    i += 1
+    j -= 1
+  end
+
+  # returns true if the the sum_of_difference equals zero, false otherwise.
+  sum_of_diff == 0 ? true : false
+
 end
 
-# Time Complexity: ?
-# Space Complexity: ?
+
+# Time Complexity:O(n^3) three nested loops, each not fixed
+# Space Complexity: O(n) because @strings's size on disk varies according to the size of the input
 def longest_prefix(strings)
```

Comment:
This method is actually O(nm) time complexity because the outer loop goes off the number of strings, n, and the inner loop goes based off the number of letters in common, m.  

Also the time complexity is based on the length of the smallest string in strings.  So O(m)